iree/turbine/kernel/ops/wave_ops.py

```diff
@@ -1057,6 +1064,53 @@ def index(self, value: dict[IndexSymbol, IndexSequence]):
         CustomOp.index.fset(self, value)
 
 
+@define_op("extract")
+@dataclass
+class Extract(CustomOp):
+    """
+    Op Rationale:
+
+    Extract is an op used to represent extracting of
+    a scalar from TKW's 1-D vector on the specified index.
+
+    This can also be viewed as indexing/slicing on the fastest dim/
+    SIMT/thread world, as opposed to the SIMD/wave world.
+    Hence, from the SIMD/wave world POV, the indexed dim
+    will be unit-size/"reduced".
+    """
+
+    register_: fx.Proxy
+    offset: IndexExpr | int
+
+    @property
+    def type(self) -> "Register":
+        # Intuition here is we are trying to extract an element
+        # from fastest dim => we reduce the fastest dim.
+        src_type = get_custom(self.register_).type
+        # Return itself if just 0-D/1-D symbolic.
+        if len(src_type.symbolic_shape) <= 1:
+            return src_type
+
+        # Typically only fastest dim has non-unit dim,
+        # but if all unit-dim get fastest/last one.
+        all_unit_dims = lambda index: all(x.size == 1 for x in index.values())
```

Comment:
Do we need this check? Can we merge this with the non_unit_dim check? if len(non_unit_dim) == 0, use fastest dim.

Reply:
done! I think :)